test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
 }
 
 
+BOOST_AUTO_TEST_CASE(call_function_returning_function)
+{
+	char const* sourceCode = R"(
+		contract test {
+			function f0() returns (uint) {
+				return 2;
+			}
+			function f1() returns (function() returns (uint)) {
+				returns f0;
+			}
+			function f2() returns (function() returns (function () returns (uint))) {
+				returns f1;
+			}
+			function f3() returns (function() returns (function () returns (function () returns (uint))))
+			{
+				returns f2;
+			}
+			function f() returns (uint) {
+				function() returns(function() returns(function() returns(function() returns(uint)))) x;
+				x = f3;
+				return x()()()();
+			}
+		}
+	)";
+
+	compileAndRun(sourceCode, 0, "C");
+	BOOST_CHECK(callContractFunction("f()") == encodeArgs(u256(2)));
+}
+
+BOOST_AUTO_TEST_CASE(array_of_functions)
+{
+	char const* sourceCode = R"(
+		contract Flow {
+			bool success;
+			function checkSuccess() returns(bool) {
+				return success;
+			}
+
+			mapping (address => function () internal returns()) stages;
+
+			function stage0() internal {
+					stages[msg.sender] = stage1;
+			}
+
+			function stage1() internal {
+					stages[msg.sender] = stage2;
+			}
+
+			function stage2() internal {
+				success = true;
+			}
+
+			function f () {
+				if (0 == steps[msg.sender])
+					stages[msg.sender] = stage0;
+				stages[msg.sender]();
+			}
+		}
+	)";
+
+	compileAndRun(sourceCode, 0, "C");
+	BOOST_CHECK(callContractFunction("checkSuccess()") == encodeArgs(false));
+	callContractFunction("f()");
```

Comment:
`BOOST_CHECK(callContractFunction("success()") == encodeArgs())` (answering also the question above)